### PR TITLE
Tests for All details api enpoint and versions

### PR DESCRIPTION
**Pull request by goyal1092**
## Description
Added tests for Geo versions all_details:

Test Cases:

* **Highlights**:Test for highlights data that come in all details api

  *  **test_highlight_without_data** : If we do not have any highlights for profile
  * **test_highlight_with_valid_data**: Just check the output for highlights where we have valid data (Happy path - Test with optimal data) 
  * **test_multiversion_highlight_with_valid_data**: Test highlights if there are multiple versions for a profilecreate multiple versions with valid data and set one version as default. 
   Check if appropriate data is returned according to the version passed to all_details endpoint
    * Check  if we get highlights for the default version(version1) when we do not pass the version in API
    * Check if we get highlights for the version we are passing in the ApiWe get highlights_v1 for version1 and highlights_v2 for version2.

  - **test_highlight_sibling_with_multiversion** : Test highlight for denominator type sibling and output should be different in case a sibling exists for a specific version.create multiple versions, create a sibling that exists for a specific version, create data for  that geography
    - Test data returned for version1 
    - Test if data returned for version2 is not the same as data returned for version1 (It tests that data from newly added sibling affected output for version2)
    
    For both versions, we will get the same highlight (highlight_v1) but the return value should be changed according to version and siblings included in calculations.

- **Profile Key Metrics**
 
   - **test_multiversion_metrics_data** : Test ky metrics if there are multiple versions for a profilecreate multiple versions with valid data and set one version as default.Check if the appropriate metric is returned according to the version passed to the all_details endpoint

     - Check  if we get metrics for the default version when we do not pass the version in API (metric_v1 is returned)

     - Check if we get metrics for the version we are passing in the ApiWe get metric_v1 for version1 and metric_v2 for version2.

  - **test_metric_sibling_with_multiversion** : Test metrics for denominator type sibling and output should be different in case a sibling exists for a specific version.

       create multiple versions, create a sibling that exists for a specific version, create data for  that geography

       - Test data returned for version1 

       - Test if data returned for version2 is not the same as data returned for version1 (It tests that data from newly added sibling affected output for version2)

       For both versions, we will get the same highlight (metric_v1) but the return value should be changed according to version and siblings included in calculations.

- **Parents data in all detailts (TestParentLayersData)**:  tests when there are two versions available for the same geo but one of its ancestor-siblings only exists in the selected version.

 - **Test for Child indicator data ( TestChildrenData)** : when there are two versions available for the same geo but one of its children only exist in the selected version

## Related Issue
https://wazimap.atlassian.net/browse/WNCM-122
https://wazimap.atlassian.net/browse/WNCM-99

## Changelog

- Moved tests for general views to separate folder and files
- Updated test_metrics_serializer.py to fix breaking tests
- Updated metrics_serializer.py to get version for denominator option

## Checklist

- [x]  🚀 is the code ready to be merged and go live?
- [x]  🛠 does it work (build) locally

### Pull Request

- [x]  📰 good title
- [x]  📝good description
- [x]  🔖 issue linked
- [x]  📖 changelog filled out

### Commits

- [x]  commits are clean
- [x]  commit messages are clean

### Code Quality

- [x]  🚧 no commented out code
- [x]  🖨 no unnecessary logging
- [x]  🎱 no magic numbers
- [x]  black was run locally (as part of the pre-commit hook)

### Testing

- [x]  ✅ added (appropriate) unit tests
- [x]  💢 edge cases in tests were considered
- [x]  ✅ ran tests locally & are passing

### CODE REVIEW

File: tests/general/base.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        self.create_boundary(self.root, self.version)
+        self.create_boundary(self.geo1, self.version)
+        self.create_boundary(self.geo2, self.version)
+
+        self.hierarchy = self.create_hierarchy(
+            self.root, self.version
+        )
+        self.profile = ProfileFactory(geography_hierarchy=self.hierarchy)
+        self.category = CategoryFactory(profile=self.profile)
+        self.location = LocationFactory(category=self.category)
```
Comment (jbothma) [inline tests/general/base.py:42]: does anything use this? It looks out of place. I know it's just moved, but if it passes without it, we can drop it.

File: tests/general/base.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        for i in range(0, count):
+            pcs.append(
+                self.create_pc(theme, i, F"pc{i}_{theme.name}")
+            )
+        return pcs
+
+    def create_dataset(self, profile, version):
+        return DatasetFactory(
+            profile=profile, version=version
+        )
```
Comment (jbothma) [inline tests/general/base.py:79]: why not just call the constructor?

File: tests/general/base.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+    def create_base_data(
+            self, version=None, geography=None,
+            hierarchy=None, profile=None,
+            create_version=True, create_boundary=True
+        ):
+
+        if create_version:
+            version = VersionFactory(name=version)
```
Comment (jbothma) [inline tests/general/base.py:158]: it isn't nice to use the same variable name for two different types. It gets confusing and introduces bugs. 
```suggestion
            version = VersionFactory(name=version_name)
```

File: tests/general/base.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        dataset = self.create_dataset(profile, version)
+        if not data:
+            data = [
+                {"gender": "male", "age": "15", "count": 1},
+                {"gender": "female", "age": "16", "count": 2},
+            ]
+
+        self.multiple_datasetdata(dataset, geography, data)
+        indicator = self.create_indicator(dataset, subindicator)
+        return indicator
```
Comment (jbothma) [inline tests/general/base.py:200]: `val` is unused, right?

File: tests/general/views/test_consolidated_profile_view.py [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+        subthemes = theme_data[0]["subthemes"]
+        assert subthemes[0]['label'] == pc3.label
+        assert subthemes[1]['label'] == pc1.label
+        assert subthemes[2]['label'] == pc4.label
+        assert subthemes[3]['label'] == pc2.label
+
+
+class TestProfileHighlightData(ConsolidatedProfileViewBase):
+
+    def test_highlight_without_data(self):
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:157]: your name sounds like there's a highlight with no data

```suggestion
    def test_zero_highlights(self):
```

File: tests/general/views/test_consolidated_profile_view.py [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+
+        dataset1 = self.create_dataset(self.profile, self.version)
+        dataset2 = self.create_dataset(self.profile, version2)
+        for dataset in [dataset1, dataset2]:
+            self.multiple_datasetdata(dataset, self.geo1, data["geo1"])
+            self.multiple_datasetdata(dataset, self.geo2, data["geo2"])
+            self.multiple_datasetdata(dataset, geo3, data["geo3"])
+            indicator = self.create_indicator(dataset, "gender")
+            self.create_highlight(
+                self.profile, indicator, "female", "highlight_v1", "sibling"
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:278]: highlight_v1 doesn't make sense when we create it for both a v1 and a v2 dataset, right?

File: tests/general/views/test_consolidated_profile_view.py [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+            profile_id=self.profile.pk,
+            geography_code=self.geo1.code,
+            data={
+                'format': 'json',
+            },
+        )
+        self.assert_http_200_ok()
+        assert response.data["profile"]["highlights"] == [
+            {'label': 'highlight_v1', 'value': 0.3333333333333333, 'method': 'sibling'}
+        ]
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:292]: I don't understand.

1. geo 1, 2, and 3 are siblings
2. 2+4+6=12
3. 2/12=0.166666666

I think at the top of this method you need to lay out the structure of hierarchy - I can't figure it out in my head.

File: tests/general/base.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+                    subindicators.append(data[key])
+                    group.subindicators = subindicators
+                    group.save()
+        return obj
+
+    def multiple_datasetdata(self, dataset, geography, datalist=[]):
+        objs = []
+        for data in datalist:
+            objs.append(self.create_datasetdata(dataset, geography, data))
+        return objs
```
Comment (jbothma) [inline tests/general/base.py:114]: this method is redundant. it's always called with a single object.

also, its name is not a verb, so I have to read all its code to see what it's doing.

I suggest we remove this method for now and just call create_datasetdata directly for now, until we need something more fancy.

File: tests/general/views/test_consolidated_profile_view.py [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+            },
+        )
+        self.assert_http_200_ok()
+        metrics = response.data["profile"]["profile_data"]["category-female"]["subcategories"]["subcategory-female"]["key_metrics"]
+        assert metrics == [
+            {'label': 'metric', 'value': 2.0, 'method': 'absolute_value'}
+        ]
+
+
+    def test_multiversion_metrics_data(self):
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:338]:
```suggestion
    def test_multiversion_key_metrics_data(self):
```

File: tests/general/views/test_consolidated_profile_view.py [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+            "v2", None, self.hierarchy, self.profile
+        )
+
+        data = [
+            {"gender": "male", "age": "15", "count": 1},
+            {"gender": "female", "age": "16", "count": 2},
+        ]
+        indicator = self.create_versioned_data(
+            version1, geography, profile, data=data
+        )
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:471]: why are we creating an indicator for a test that doesn't care about indicators or indicator data?

File: tests/general/views/test_consolidated_profile_view.py [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+        metrics = response.data["profile"]["profile_data"]["category-female"]["subcategories"]["subcategory-female"]["key_metrics"]
+        assert metrics == [
+            {'label': 'metric', 'value': 0.25, 'method': 'sibling'}
+        ]
+
+
+class TestParentLayersData(ConsolidatedProfileViewBase):
+
+    def test_version_parent_and_siblings(self):
+        version1, geography, hierarchy, profile = self.create_base_data(
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:461]: Docstrings of what these tests are asserting would be extremely helpful. I even wrote most of the text already :D

Screenshot_2022-02-10_17-06-22


```suggestion
    def test_version_parent_and_siblings(self):
        """Ensure that..."""
        version1, geography, hierarchy, profile = self.create_base_data(
```

Is this one testing

> when there are two versions available for the same geo but one of its ancestor-siblings only exist in the selected version, that ancestor-sibling should only be included in the all_details response for the version where it exists

?

File: tests/general/views/test_consolidated_profile_view.py [hunk 12]
```diff
@@ -0,0 +1,566 @@
...
+        assert geo_data["parents"][0]["code"] == self.root.code
+        assert geo_data["parents"][1]["code"] == self.geo1.code
+
+        # Assert parent & siblings and parent siblings
+        parent_layers = response.data["parent_layers"]
+        assert len(parent_layers) == 2
+        assert len(parent_layers[0]["features"]) == 1
+        assert parent_layers[0]["features"][0]["properties"]["code"] == self.geo1.code
+        assert len(parent_layers[1]["features"]) == 1
+        assert parent_layers[1]["features"][0]["properties"]["code"] == geography.code
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:496]: again, the hierarchy of geos in each version is critical to quickly be able to understand this test, and to be able to edit it correctly in future.

File: tests/general/views/test_consolidated_profile_view.py [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+
+        # Add boundary for version & parent sibling
+        self.create_boundary(self.geo2, version1)
+        # Add sibling but with same version
+        self.create_boundary(self.dc1, version1)
+
+        self.tearDown()
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:503]: wait what? we're tearing down inside the test? That doesn't look like a good idea!

Should this be two tests?

File: tests/general/views/test_consolidated_profile_view.py (hunk 12 above)
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:496]: I like the assertion on length == 1 and then using index 0. neither one of these assertions are sufficient, but together they're awesome.

File: tests/general/views/test_consolidated_profile_view.py (hunk 13 above)
Comment (jbothma) [reply tests/general/views/test_consolidated_profile_view.py:503]: I'll have to review this once this is fixed.

File: tests/general/views/test_consolidated_profile_view.py [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+            "v2", None, self.hierarchy, self.profile
+        )
+
+        data = [
+            {"gender": "male", "age": "15", "count": 1},
+            {"gender": "female", "age": "16", "count": 2},
+        ]
+        indicator = self.create_versioned_data(
+            version1, geography, profile, data=data
+        )
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:537]: do we need this?

File: tests/general/views/test_consolidated_profile_view.py [hunk 15]
```diff
@@ -0,0 +1,566 @@
...
+        assert len(parent_layers[1]["features"]) == 2
+        assert parent_layers[1]["features"][0]["properties"]["code"] == self.dc1.code
+        assert parent_layers[1]["features"][1]["properties"]["code"] == geography.code
+
+class TestChildrenData(ConsolidatedProfileViewBase):
+
+    def test_version_parent_and_siblings(self):
+        version1, geography, hierarchy, profile = self.create_base_data(
+            "v2", None, self.hierarchy, self.profile
+        )
```
Comment (jbothma) [inline tests/general/views/test_consolidated_profile_view.py:529]: please document the geo hierarchy for each version in a comment for this test too.

Comment (jbothma) [reply tests/general/views/test_consolidated_profile_view.py:529]: an alertnative to documenting the geo levels might be to name them in a way that conveys this, e.g. geo1a for root, geo1a2a and geo1a2b, and geo1a2b3a and so on.... or not :P

File: tests/general/base.py (hunk 1 above)
Comment (goyal1092) [reply tests/general/base.py:42]: Yes we are need locations for themes-count api. Theme is excluded if there is no location in it.

File: tests/general/base.py (hunk 2 above)
Comment (goyal1092) [reply tests/general/base.py:79]: `create_dataset ` is being called by `create_versioned_data` that is creating versioned data for possible a diff profile or version each time it is called.

File: tests/general/base.py (hunk 4 above)
Comment (goyal1092) [reply tests/general/base.py:200]: yes, Removed it 👍

File: tests/general/views/test_consolidated_profile_view.py (hunk 6 above)
Comment (goyal1092) [reply tests/general/views/test_consolidated_profile_view.py:278]: Actually, it makes sense to have the same label of highlight for a different version.
```
dataset of v1 -> indicator of v1 -> highlight
 dataset of v2 -> indicator of v2 -> highlight
```
Both of the highlights should have the same label, So in the frontend, only values of highlight change according to version but the label remains same. 
This was my knowledge of how we handled highlight for diff versions at that point. Not sure if it has changed or I misunderstood how it worked on frontend
 
 But will change it in tests for more readability.

File: tests/general/views/test_consolidated_profile_view.py (hunk 7 above)
Comment (goyal1092) [reply tests/general/views/test_consolidated_profile_view.py:292]: removing loop 👍

File: tests/general/base.py [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+                        dataset=dataset, name="gender", subindicators=[],
+                        can_aggregate=True, can_filter=True
+                    )
+
+                if data[key] not in group.subindicators:
+                    subindicators = group.subindicators
+                    subindicators.append(data[key])
+                    group.subindicators = subindicators
+                    group.save()
+        return obj
```
[the remark this answers is not in the document]
Comment (goyal1092) [reply tests/general/base.py]: Changed to use datasets/task/upload/process_csv to create datasetdata

File: tests/general/base.py (hunk 3 above)
Comment (goyal1092) [reply tests/general/base.py]: Done

File: tests/general/views/test_consolidated_profile_view.py (hunk 5 above)
Comment (goyal1092) [reply tests/general/views/test_consolidated_profile_view.py]: Done

File: tests/general/views/test_consolidated_profile_view.py (hunk 6 above)
Comment (goyal1092) [reply tests/general/views/test_consolidated_profile_view.py]: Done

File: tests/general/views/test_consolidated_profile_view.py (hunk 13 above)
Comment (goyal1092) [reply tests/general/views/test_consolidated_profile_view.py]: Converted this test case to be 2 tests

File: tests/general/views/test_consolidated_profile_view.py (hunk 14 above)
Comment (goyal1092) [reply tests/general/views/test_consolidated_profile_view.py]: removed it

File: tests/general/views/test_task_details.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+@pytest.mark.focus
+@pytest.mark.django_db
+def test_task_status(tp, tp_api):
+    reversed_url = tp.reverse(
+        "task_status",
+        task_id="1234",
+    )
+    response = tp_api.client.get(reversed_url, format="json")
+
+    assert response.status_code == 200
```
[the remark this answers is not in the document]
Comment (goyal1092) [reply tests/general/views/test_task_details.py:16]: It's coming from Django test plus
https://github.com/revsys/django-test-plus#pytest-usage

**Comment by goyal1092** [comment]
Ready for review @jbothma

**Review by michaelglenister — COMMENTED** [review]
It looks like all previous review comments have been attended to. The consolidated profile view tests are still complex but now much more understandable

File: tests/general/base.py [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+from tests.datasets.tasks.test_process_uploaded_file import create_datasetfile
+
+
+class ConsolidatedProfileViewBase(APITestCase):
+
+    def setUp(self):
+        """
+        Create Data for default_version
+        """
+        self.default_version = VersionFactory(name="default_version")
```
Comment (jbothma) [inline tests/general/base.py:31]:
```suggestion
        self.default_version = VersionFactory(name="v1")
```

File: tests/general/base.py [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+        self.geo2 = self.root.add_child(code="CHILD2", level="province")
+        self.geo3 = self.root.add_child(code="CHILD3", level="province")
+
+        self.create_boundary(self.root, self.default_version)
+        self.create_boundary(self.geo1, self.default_version)
+        self.create_boundary(self.geo2, self.default_version)
+
+        # Create hierarchy
+        self.hierarchy = self.create_hierarchy(
+            self.root, self.default_version, versions=["default_version"]
```
Comment (jbothma) [inline tests/general/base.py:45]:
```suggestion
            self.root, self.default_version, versions=["v1"]
```

File: tests/general/base.py (hunk 18 above)
Comment (jbothma) [reply tests/general/base.py:31]: I think it's worth not conflating GeographyHierarchy.config.default_version with a version named "default_version".

When I see a version named v1, I assume that's the first (and only) version.

I think if tests that use a second Version create a version named v2, it should be quite clear that that's the second version in the system.

File: tests/general/views/test_consolidated_profile_view.py (hunk 6 above)
Comment (jbothma) [reply tests/general/views/test_consolidated_profile_view.py]: having the same name was fine, the issue was that the name had the suffix _v1 for both the v1 and the v2 dataset.

**Review by jbothma — APPROVED** [review]
